src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+                }
+            }
+            return (methodSymbol != null && methodSymbol.GetParameters().Any(), hasOptionalParameter);
+        }
+
+        private (string title, ImmutableArray<CodeAction> actions) AddActions(SemanticDocument semanticDocument, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            var actionsBuilder = new ArrayBuilder<CodeAction>();
+            var (isParameterized, hasOptionalParameter) = ExpressionWithinParameterizedMethod(semanticDocument, expression, cancellationToken);
+            if (isParameterized)
+            {
+                actionsBuilder.Add(new IntroduceParameterCodeAction(semanticDocument, (TService)this, expression, false, false, false));
+                actionsBuilder.Add(new IntroduceParameterCodeAction(semanticDocument, (TService)this, expression, true, false, false));
+
+                if (!hasOptionalParameter)
+                {
+                    actionsBuilder.Add(new IntroduceParameterCodeAction(semanticDocument, (TService)this, expression, false, true, false));
+                    actionsBuilder.Add(new IntroduceParameterCodeAction(semanticDocument, (TService)this, expression, true, true, false));
+
+                    actionsBuilder.Add(new IntroduceParameterCodeAction(semanticDocument, (TService)this, expression, false, false, true));
+                    actionsBuilder.Add(new IntroduceParameterCodeAction(semanticDocument, (TService)this, expression, true, false, true));
+                }
+            }
+
+            return (FeaturesResources.Introduce_parameter, actionsBuilder.ToImmutable());
+        }
+
+        protected static ISet<TExpressionSyntax> FindMatches(SemanticDocument originalDocument,
+                                                             TExpressionSyntax expressionInOriginal,
+                                                             SemanticDocument currentDocument,
+                                                             SyntaxNode withinNodeInCurrent,
+                                                             bool allOccurrences,
+                                                             CancellationToken cancellationToken)
+        {
+            var syntaxFacts = currentDocument.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var originalSemanticModel = originalDocument.SemanticModel;
+            var currentSemanticModel = currentDocument.SemanticModel;
+
+            var result = new HashSet<TExpressionSyntax>();
+            var matches = from nodeInCurrent in withinNodeInCurrent.DescendantNodesAndSelf().OfType<TExpressionSyntax>()
+                          where NodeMatchesExpression(originalSemanticModel, currentSemanticModel, expressionInOriginal, nodeInCurrent, allOccurrences, cancellationToken)
+                          select nodeInCurrent;
+            result.AddRange(matches.OfType<TExpressionSyntax>());
+
+            return result;
```

Comment:
what's going on here?

Reply:
it's going through all the expressionsyntax nodes in the container and determining if they match the original highlighted expression, then it's adding them into a hashset